modules/ip/control/nexthop.c

```diff
@@ -10,71 +10,58 @@
 #include <gr_log.h>
 #include <gr_net_types.h>
 #include <gr_queue.h>
+#include <gr_stb_ds.h>
 
 #include <event2/event.h>
 #include <rte_errno.h>
-#include <rte_ethdev.h>
-#include <rte_hash.h>
-#include <rte_malloc.h>
+#include <rte_ether.h>
+#include <rte_mempool.h>
 
 #include <errno.h>
 #include <stdio.h>
 #include <string.h>
 #include <sys/queue.h>
 
-static struct nexthop *nh_array;
-static struct rte_hash *nh_hash;
+static struct rte_mempool *nh_pool;
 
-struct nexthop_key {
-	ip4_addr_t ip;
-	// XXX: Using uint16_t to hold vrf_id causes the compiler to add 2 bytes
-	// padding at the end of the structure. When the structure is
-	// initialized on the stack, the padding bytes have undetermined
-	// contents.
-	//
-	// This structure is used to compute a hash key. In order to get
-	// deterministic results, use uint32_t to store the vrf_id so that the
-	// compiler does not insert any padding.
-	uint32_t vrf_id;
-};
-
-struct nexthop *ip4_nexthop_get(uint32_t idx) {
-	return &nh_array[idx];
-}
-
-int ip4_nexthop_lookup(uint16_t vrf_id, ip4_addr_t ip, uint32_t *idx, struct nexthop **nh) {
-	struct nexthop_key key = {ip, vrf_id};
-	int32_t nh_idx;
+struct nexthop *ip4_nexthop_new(uint16_t vrf_id, uint16_t iface_id, ip4_addr_t ip) {
+	struct nexthop *nh;
+	void *data;
+	int ret;
 
-	if ((nh_idx = rte_hash_lookup(nh_hash, &key)) < 0)
-		return errno_set(-nh_idx);
+	if ((ret = rte_mempool_get(nh_pool, &data)) < 0)
+		return errno_set_null(-ret);
 
-	*idx = nh_idx;
-	*nh = &nh_array[nh_idx];
+	nh = data;
+	nh->vrf_id = vrf_id;
+	nh->iface_id = iface_id;
+	nh->ip = ip;
```

Comment:
`nh->ref_count = 1` ?

Reply:
I thought about it. The issue is that ip4_route_insert() also increments the ref count. It will result in nexthop leaking. I don't have an elegant solution to this.